File: Harbor.Python/Tool/ConvertUpper.py
from pyverilog.vparser.parser import parse
import pyverilog.vparser.ast as vast
from pyverilog.ast_code_generator.codegen import ASTCodeGenerator

VDD = 'DVDD'
VSS = 'DVSS'

def get_ports(node, port_lists):
    for i in node.children():
        i_name = i.__class__.__name__
        if i_name == "ModuleDef":
            module_name = i.name
            # print(module_name)
            portlist = i.portlist
            portlist = list(portlist.ports)
            real_port_list = []
            for p in portlist:
                if p.__class__.__name__ == 'Ioport': # 带有input或者output声明的
                    # print(p.first.name)
                    real_port_list.append(p.first.name)
                else: # 不带任何声明
                    # print(p.name)
                    real_port_list.append(p.name)

            port_lists[module_name] = real_port_list

        get_ports(i, port_lists)

def convert_ports_to_upper_in_function(node):
    for i in node.children():
        if i.__class__.__name__ == 'Input':
            i.name = i.name.upper()

        convert_ports_to_upper_in_function(i)


def convert_identifier_to_upper_in_module(src_ports, node):
    for i in node.children():
        i_name = i.__class__.__name__
        if i_name == "Identifier":
            if i.name in src_ports:
                i.name = i.name.upper()
            
        # elif i_name == "PortArg":
        #     i.portname = i.portname.upper()
        # elif i_name == "Wire":
        #     i.name = i.name.upper()
        # elif i_name == "Reg":
        #     i.name = i.name.upper()
        # elif i_name == "Integer":
        #     i.name = i.name.upper()
        # elif i_name == "Real":
        #     i.name = i.name.upper()
        # elif i_name == "Genvar":
        #     i.name = i.name.upper()

        convert_identifier_to_upper_in_module(src_ports, i)

def convert_ports_to_upper(src_ports, netlist_ports, node):
    for i in node.children():
        i_name = i.__class__.__name__
        if i_name == "ModuleDef":
            # 仅修改模块的Port, 并且根据netlist调整Port顺序
            module_name = i.name
            if module_name == top:
                # i.show()
                portlist = i.portlist
                portlist = list(portlist.ports)
                # print(portlist)
                new_order_port_list = []
                if module_name in netlist_ports: # 如果netlist含有这个模块,则调整顺序并大写
                    for target_ordered_port in netlist_ports[module_name]:
                        for origin_ordered_port in portlist:
                            origin_ordered_port_name = ''
                            if origin_ordered_port.__class__.__name__ == "Ioport":
                                origin_ordered_port_name = origin_ordered_port.first.name.upper()
                                origin_ordered_port.first.name = origin_ordered_port_name
                            else:
                                origin_ordered_port_name = origin_ordered_port.name.upper()
                                origin_ordered_port.name = origin_ordered_port_name
                            
                            if origin_ordered_port_name == target_ordered_port:
                                new_order_port_list.append(origin_ordered_port)
                else: # 如果netlist不不包含这个模块,则仅大写
                    for p in portlist:
                        if p.__class__.__name__ == "Ioport":
                            p.first.name = p.first.name.upper()
                        else:
                            p.name == p.name.upper()
                        new_order_port_list.append(p)

                
                dvdd = vast.Ioport(first = vast.Input(name = VDD), lineno=-1)
                dvss = vast.Ioport(first = vast.Input(name = VSS), lineno=-1)
                new_order_port_list.append(dvdd)
                new_order_port_list.append(dvss)
                i.portlist.ports = tuple(new_order_port_list)

                convert_identifier_to_upper_in_module(src_ports[top], i)

        convert_ports_to_upper(src_ports, netlist_ports, i)


src_ast, src_directives = parse([source], preprocess_include=[], preprocess_define=[])
netlist_ast, netlist_directives = parse([netlist], preprocess_include=[], preprocess_define=[])

src_ports = {}
get_ports(src_ast, src_ports)
netlist_ports = {}
get_ports(netlist_ast, netlist_ports)

convert_ports_to_upper(src_ports, netlist_ports, src_ast)

codegen = ASTCodeGenerator()
rslt = codegen.visit(src_ast)

File: Harbor.Python/Tool/ConvertUpper.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using Python.Runtime;

namespace Harbor.Python.Tool
{
    public static class ConvertUpper
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="top"></param>
        /// <param name="source">RTL源代码</param>
        /// <param name="netlist">LAYOUT后输出的网表,将会提取其端口顺序</param>
        /// <param name="output">输出文件名称,输出文件的端口将会全部大写,并且添加DVDD和DVSS占位,并且端口顺序和netlist.v相同</param>
        public static void Run(string top, string source, string netlist, string output, string workingDirectory)
        {
            PythonHelper.SetEnvironment(workingDirectory, () =>
            {
                dynamic pyverilog = Py.Import("pyverilog");
                dynamic parser = Py.Import("pyverilog.vparser.parser");
                dynamic vast = pyverilog.vparser.ast;
                dynamic codegen = Py.Import("pyverilog.ast_code_generator.codegen");

                void GetPorts(dynamic node, Dictionary<string, List<string>> portLists)
                {
                    foreach (var i in node.children())
                    {
                        var iName = i.__class__.__name__.As<string>();
                        if (iName == "ModuleDef")
                        {
                            var moduleName = i.name.As<string>();
                            var realPortList = new List<string>();
                            foreach (var p in i.portlist.ports)
                            {
                                realPortList.Add(p.__class__.__name__.As<string>() == "Ioport"
                                    ? (string) p.first.name
                                    : (string) p.name);
                            }

                            portLists.Add(moduleName, realPortList);
                        }

                        GetPorts(i, portLists);
                    }
                }

                void ConvertPortsToUpperInFunction(dynamic node)
                {
                    foreach (var i in node.children())
                    {
                        if (i.__class__.__name__.As<string>() == "Input")
                        {
                            string n = i.name.ToString();
                            i.name = n.ToUpper();
                        }

                        ConvertPortsToUpperInFunction(i);
                    }
                }

                void ConvertIdentifierToUpperInModule(List<string> srcps, dynamic node)
                {
                    foreach (var i in node.children())
                    {
                        if (i.__class__.__name__.As<string>() == "Identifier")
                        {
                            string n = i.name.ToString();
                            if (srcps.Contains(n))
                            {
                                i.name = n.ToUpper();
                            }
                        }

                        ConvertIdentifierToUpperInModule(srcps, i);
                    }
                }

                void ConvertPortsToUpper(Dictionary<string, List<string>> srcps,
                    Dictionary<string, List<string>> netlistps, dynamic node)
                {
                    foreach (var i in node.children())
                    {
                        var iName = i.__class__.__name__.As<string>();
                        if (iName == "ModuleDef")
                        {
                            //仅修改模块的Port, 并且根据netlist调整Port顺序
                            var moduleName = i.name.As<string>();
                            if (moduleName == top)
                            {
                                dynamic portList = i.portlist.ports;
                                var newOrderPortList = new List<PyObject>();
                                if (netlistps.ContainsKey(moduleName)) // 如果netlist含有这个模块,则调整顺序并大写
                                {
                                    foreach (string targetOrderedPort in netlistps[moduleName])
                                    {
                                        foreach (var originOrderedPort in portList)
                                        {
                                            string originOrderedPortName;
                                            if (originOrderedPort.__class__.__name__.As<string>() == "Ioport")
                                            {
                                                originOrderedPortName = originOrderedPort.first.name.As<string>();
                                                originOrderedPortName = originOrderedPortName.ToUpper();
                                                originOrderedPort.first.name = originOrderedPortName;
                                            }
                                            else
                                            {
                                                originOrderedPortName = originOrderedPort.name.As<string>();
                                                originOrderedPortName = originOrderedPortName.ToUpper();
                                                originOrderedPort.first.name = originOrderedPortName;
                                            }

                                            if (originOrderedPortName == targetOrderedPort)
                                            {
                                                newOrderPortList.Add(originOrderedPort);
                                            }
                                        }
                                    }
                                }
                                else //如果netlist不不包含这个模块,则仅大写
                                {
                                    foreach (var p in portList)
                                    {
                                        if (p.__class__.__name__.As<string>() == "Ioport")
                                        {
                                            string n = p.first.name.As<string>();
                                            p.first.name = n.ToUpper();
                                        }
                                        else
                                        {
                                            string n = p.name.As<string>();
                                            p.first.name = n.ToUpper();
                                        }

                                        newOrderPortList.Add(p);
                                    }
                                }

                                dynamic dvdd = vast.Ioport(first: vast.Input(name: "DVDD"), lineno: -1);
                                dynamic dvss = vast.Ioport(first: vast.Input(name: "DVSS"), lineno: -1);

                                newOrderPortList.Add(dvdd);
                                newOrderPortList.Add(dvss);
                                i.portlist.ports = newOrderPortList;

                                ConvertIdentifierToUpperInModule(srcps[top], i);
                            }
                        }

                        ConvertPortsToUpper(srcps, netlistps, i);
                    }
                }

                dynamic srcTuple = parser.parse(new List<string> {source});
                dynamic srcAst = srcTuple[0];
                dynamic netlistTuple = parser.parse(new List<string> {netlist});
                dynamic netlistAst = netlistTuple[0];

                var srcPorts = new Dictionary<string, List<string>>();
                var netlistPorts = new Dictionary<string, List<string>>();
                GetPorts(srcAst, srcPorts);
                GetPorts(netlistAst, netlistPorts);
                ConvertPortsToUpper(srcPorts, netlistPorts, srcAst);

                dynamic codegenI = codegen.ASTCodeGenerator();
                string rslt = codegenI.visit(srcAst).As<string>();

                File.WriteAllText(output, PythonHelper.Banner + DateTime.Now + Environment.NewLine + rslt);
            });
        }

        public static void Run2(string top, string source, string netlist, string output, string workingDirectory)
        {
            var className = MethodBase.GetCurrentMethod()?.DeclaringType?.FullName;
            var code = PythonHelper.GetCodeFromResource($"{className}.py");

            string rslt = "";
            PythonHelper.SetEnvironment(workingDirectory, () =>
            {
                using var scope = Py.CreateScope();
                scope.Set("source", source);
                scope.Set("netlist", netlist);
                scope.Set("top", top);
                scope.Exec(code);
                rslt = scope.Get<string>("rslt");
            });
            File.WriteAllText(output, PythonHelper.Banner + DateTime.Now + Environment.NewLine + rslt, new UTF8Encoding(false));
        }
    }
}
